realm/src/androidTest/java/io/realm/NotificationsTest.java

```diff
@@ -347,4 +388,76 @@ public void onChange() {
         realm.commitTransaction();
         assertTrue(success.get());
     }
+
+    public void testAddRemoveListenerConcurrency() {
+        final AtomicInteger counter1 = new AtomicInteger(0);
+        final AtomicInteger counter2 = new AtomicInteger(0);
+        final AtomicInteger counter3 = new AtomicInteger(0);
+
+        // At least we need 2 listeners existing in the list to make sure
+        // the iterator.next get called
+        final RealmChangeListener listener1 = new RealmChangeListener() {
+            @Override
+            public void onChange() {
+                counter1.incrementAndGet();
+            }
+        };
+
+        final RealmChangeListener listener2 = new RealmChangeListener() {
+            @Override
+            public void onChange() {
+                counter2.incrementAndGet();
+                realm.addChangeListener(listener1);
+            }
+        };
+
+        RealmChangeListener listener3 = new RealmChangeListener() {
+            @Override
+            public void onChange() {
+                counter3.incrementAndGet();
+                realm.removeChangeListener(this);
+            }
+        };
+
+        realm = Realm.getInstance(getContext());
```

Comment:
Some comments here about the intermediates step would be nice, as the interaction between counters and listeners are very subtle.